### ISSUE TITLE
discussion: move asserts to Deno.assert

### ISSUE THREAD

**Issue by bartlomieju** [denoland/deno#4090]
With recent move of testing functions from `std/testing/mod.ts` to `Deno.test()` it seems silly to require importing assertions from standard library to write tests.

My proposition is to move methods available in `std/testing/asserts.ts` under `Deno.assert` namespace; it would make testing experience in Deno completely built-in.

Example test file before:
```ts
import { assertEquals } from "https://deno.land/std/testing/asserts.ts";

Deno.test(function testSomething() {
   const a = "foobar";
   const b = "fizzbuz";
   assertEquals(a, b);
});
```

Example test file after:
```ts
Deno.test(function testSomething() {
   const a = "foobar";
   const b = "fizzbuz";
   Deno.assert.equals(a, b);
});
```

Note: `std/testing/asserts.ts` depends on `std/testing/diff.ts` and `std/testing/format.ts`. Former is needed for REPL anyway (#3760) latter can potentially be removed (#4023, #3165)

**Comment by kitsonk** [comment]
Personal opinion, we should only move `assert` and `AssertionError` class, instead of dragging in a whole load of stuff.  Those can be built upon downstream, but it really should be up to the test harness to format an assertion error when it throws.

I suggested `AssertionError` a way back, but I didn't fully articulate what it should be, and now that I had more experience in this space in a side project, I believe `AssertionError` should be:

```ts
class AssertionError extends Error {
  expected?: any;
  actual?: any;
  showDiff?: boolean;

  constructor(message: string) {
    super(message);
    this.name = "AssertionError";
  }
}
```

And the display logic though be figured out by whomever catches the `AssertionError` instead of the display logic being in the assertion.  This would make it compatible with other test harnesses (like Jest for example).  While Jest doesn't respect the `showDiff` flag, there are other libraries that do.  This would move all the display logic to `Deno.test` or whatever other harness people used.

**Comment by bartlomieju** [comment]
> Personal opinion, we should only move assert and AssertionError class, instead of dragging in a whole load of stuff. Those can be built upon downstream, but it really should be up to the test harness to format an assertion error when it throws.

I completely agree with you that display logic should be handled by test framework itself, but personally I think that methods available in `asserts.ts` are very useful - they cover **a lot** of assertions you might want to use (obviously, we should go through all of them one-by-one and vet which ones should actually make it to `Deno.asserts`). Side effect is that it'd be possible to polyfill Node's `assert` easily. 

> And the display logic though be figured out by whomever catches the AssertionError instead of the display logic being in the assertion. This would make it compatible with other test harnesses (like Jest for example). While Jest doesn't respect the showDiff flag, there are other libraries that do. This would move all the display logic to Deno.test or whatever other harness people used.

In case of Deno; once #3165 lands; there won't be much bloat to have pretty printed assertion errors, it's `testing/format.ts` that I really wouldn't want to move to `cli/js/. 

As described earlier `diff` method would be useful in REPL anyway; and in case of `std/colors`, it's already used in `cli/js/` so that won't make any difference.

**Comment by kitsonk** [comment]
The swmi-concen is top level bloat of `Deno` stuff. At one point we decorated `assert` with properties so you could do `assert()` or `assert.equals()`. Personally I would like us to do that again (along with refactoring the display logic to `Deno.test()`.

I haven't looked in depth of the logic of `diff`. It would be really good if it was really general purpose. I built one (which also provided a `patch()` which could apply mutations as well. I can dig it out.

**Comment by bartlomieju** [comment]
Closing in favor of https://github.com/denoland/deno/issues/5684
